### PR TITLE
keycloak packer + pulumi first round 

### LINKED ISSUE #1330: Create an AMI build for Keycloak server
# User Story
- As a platform engineer I want a reliable and repeatable way of building a deployable artifact for running Keycloak as a production service.

# Description/Context
There is an available Docker image for running Keycloak that we can use as-is in part of a docker-compose deployment. We will also want to use Traefik as the reverse proxy and deploy as a multi-instance service behind a load balancer.

# Acceptance Criteria
- [ ] There is a Packer template and PyInfra code and configuration for building a deployable Keycloak instance.

### PR DESCRIPTION
**Pull request by Ardiea**
## Description
This pull request has the configuration for a basic keycloak AMI build + the basic structure for a keycloak environment provisioned with pulumi. It is not ready for production. 

Still need to 
- write structure around a keycloak.conf file and mount that into the docker container 
  - (and figure out where in the container it goes).
- Need to add custom providers we need either via a custom Dockerfile or with volume mounts.
- Probably a ton of other stuff too.

Further helpful documentation
- https://www.keycloak.org/server/containers
- https://www.keycloak.org/guides#server

## Motivation and Context
Closes #1330 

## How Has This Been Tested?
Provisioned a CI environment that is stable and does not boot-loop.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (improves on existing behavior)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project. (Did you install and run the pre-commit hooks?)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Ardiea** [comment]
Partially addresses #1332

File: src/bilder/images/keycloak/files/docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+services:
+  keycloak:
+    image: quay.io/keycloak/keycloak:${VERSION}
+    command:
+    - "start-dev"
+    env_file:
+    - ./.env
+    labels:
+    - "traefik.enable=true"
+    - "traefik.http.routers.keycloak.rule=Host(`${KEYCLOAK_HOST}`)"
```
Comment (blarghmatey) [inline src/bilder/images/keycloak/files/docker-compose.yaml:11]: Should this be `KC_HOSTNAME`?

File: src/ol_infrastructure/applications/keycloak/__main__.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+)
+from ol_infrastructure.lib.aws.ec2_helper import InstanceTypes, default_egress_args
+from ol_infrastructure.lib.consul import consul_key_helper, get_consul_provider
+from ol_infrastructure.lib.ol_types import AWSBase
+from ol_infrastructure.lib.pulumi_helper import parse_stack
+from ol_infrastructure.lib.stack_defaults import defaults
+from ol_infrastructure.lib.vault import setup_vault_provider
+
+# Configuration items and initialziations
+if Config("vault_server").get("env_namespace"):
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/keycloak/__main__.py:45]: We can remove these conditionals. That was largely a "feature flag" while we got this automatic provider setup working and migrated over.

File: src/ol_infrastructure/applications/keycloak/__main__.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+keycloak_database_security_group = ec2.SecurityGroup(
+    f"keycloak-database-security-group-{env_name}",
+    name=f"keycloak-database-{target_vpc_name}-{env_name}",
+    description="Access control for the keycloak database",
+    ingress=[
+        ec2.SecurityGroupIngressArgs(
+            security_groups=[
+                keycloak_server_security_group.id,
+                consul_stack.require_output("security_groups")["consul_server"],
+                vault_stack.require_output("vault_server")["security_group"],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/keycloak/__main__.py:140]: We'll likely also want to allow access from the data integrator SG for being able to incorporate user info into the data platform.

File: src/ol_infrastructure/applications/keycloak/__main__.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    port=db_port,
+    meta={
+        "external-node": True,
+        "external-probe": True,
+    },
+    checks=[
+        ServiceCheckArgs(
+            check_id="keycloak-instance-db",
+            interval="10s",
+            name="keycloak-instance-db",
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/keycloak/__main__.py:206]: I'd probably shorten this to just `keycloak-db`

File: src/ol_infrastructure/applications/keycloak/__main__.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+    ],
+    most_recent=True,
+    owners=[str(aws_account.id)],
+)
+
+block_device_mappings = [BlockDeviceMapping()]
+
+keycloak_lb_config = OLLoadBalancerConfig(
+    enable_insecure_http=True,
+    listener_cert_domain="video.odl.mit.edu",
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/keycloak/__main__.py:253]: Copy :spaghetti: ?

File: src/ol_infrastructure/applications/keycloak/keycloak_server_policy.hcl [hunk 6]
```diff
@@ -0,0 +1,17 @@
+path "secret-operations/*" {
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/keycloak/keycloak_server_policy.hcl:1]: This seems a bit broad? Maybe scope it specifically to a path that holds the relevant Keycloak secrets?

Comment (Ardiea) [reply src/ol_infrastructure/applications/keycloak/keycloak_server_policy.hcl:1]: Yeah I think this was laziness on my part and not knowing what I needed from secret-operations/ yet. I think at the moment we just need the wildcard cert and I did make a secret-keycloak/* mount explicitly for this.

**Review by blarghmatey — APPROVED** [review]
